core/src/test/java/org/apache/iceberg/avro/AvroTestHelpers.java

```diff
@@ -126,9 +139,18 @@ private static void assertEquals(Type type, Object expected, Object actual) {
         assertThat(actual).as("Primitive value should be equal to expected").isEqualTo(expected);
         break;
       case STRUCT:
-        assertThat(expected).as("Expected should be a Record").isInstanceOf(Record.class);
-        assertThat(actual).as("Actual should be a Record").isInstanceOf(Record.class);
-        assertEquals(type.asStructType(), (Record) expected, (Record) actual);
+        assertThat(expected)
```

Comment:
The problem with this approach is that it allows the struct types to be mixed. If for some reason the reader produced a top-level Avro `GenericData.Record` but nested structs were `StructLike`, this would allow it as long as the expected value passed in matches. Unfortunately, it's not uncommon for people to submit PRs where the expected values are set to whatever makes tests pass.

Given the amount of code here, I'd prefer to either add a boolean flag to the `private` methods or use a separate set of methods so that the record class is validated using the same logic throughout.

Reply:
This is tricky. The maps and lists also can have internal members as struct (recursive calls). So, I didn't use boolean flag here as non-private methods also need to pass the boolean.
 
But I can understand the concern about mixed types in struct. I have added a `InternalTestHelper` now with code duplication. It can be used for parquet internal test also.